### PR TITLE
refactor `fn bootstrap::builder::Builder::compiler_for` logic

### LINKED ISSUE #109286: bootstrap: Error: Unable to find the stamp file
With `profile = "tools"`, and a empty `build/` folder, and running `./x test ./tests/rustdoc-json/`, I get the following error

```
Building test helpers
running: "cc" "-O0" "-ffunction-sections" "-fdata-sections" "-fPIC" "-m64" "-o" "/home/alona/dev/rust/rust/build/x86_64-unknown-linux-gnu/native/rust-test-helpers/rust_test_helpers.o" "-c" "/home/alona/dev/rust/rust/tests/auxiliary/rust_test_helpers.c"
exit status: 0
running: ZERO_AR_DATE="1" "ar" "cq" "/home/alona/dev/rust/rust/build/x86_64-unknown-linux-gnu/native/rust-test-helpers/librust_test_helpers.a" "/home/alona/dev/rust/rust/build/x86_64-unknown-linux-gnu/native/rust-test-helpers/rust_test_helpers.o"
exit status: 0
running: "ar" "s" "/home/alona/dev/rust/rust/build/x86_64-unknown-linux-gnu/native/rust-test-helpers/librust_test_helpers.a"
exit status: 0
Uplifting library (stage2 -> stage0)
Error: Unable to find the stamp file /home/alona/dev/rust/rust/build/x86_64-unknown-linux-gnu/stage2-std/x86_64-unknown-linux-gnu/release/.libstd.stamp, did you try to keep a nonexistent build stage?
Build completed unsuccessfully in 0:01:17
```

On 03b01c5bec658081605ab078ad3fbcdb6b30f6c2, x86-64, linux.

<details>

<summary> Full Log </summary>

CC @jyn514 
```
alona@Ashtabula:~/dev/rust/rust$ rm -rf build/
alona@Ashtabula:~/dev/rust/rust$ cat config.toml
# Includes one of the default files in src/bootstrap/defaults
profile = "tools"
changelog-seen = 2
alona@Ashtabula:~/dev/rust/rust$ ./x test ./tests/rustdoc-json/
downloading https://static.rust-lang.org/dist/2023-01-25/rust-std-beta-x86_64-unknown-linux-gnu.tar.xz
##################################################################################################################################################### 100.0%
extracting /home/alona/dev/rust/rust/build/cache/2023-01-25/rust-std-beta-x86_64-unknown-linux-gnu.tar.xz
downloading https://static.rust-lang.org/dist/2023-01-25/rustc-beta-x86_64-unknown-linux-gnu.tar.xz
##################################################################################################################################################### 100.0%
extracting /home/alona/dev/rust/rust/build/cache/2023-01-25/rustc-beta-x86_64-unknown-linux-gnu.tar.xz
downloading https://static.rust-lang.org/dist/2023-01-25/cargo-beta-x86_64-unknown-linux-gnu.tar.xz
##################################################################################################################################################### 100.0%
extracting /home/alona/dev/rust/rust/build/cache/2023-01-25/cargo-beta-x86_64-unknown-linux-gnu.tar.xz
Building bootstrap
   Compiling libc v0.2.137
   Compiling memchr v2.5.0
   Compiling cfg-if v1.0.0
   Compiling typenum v1.15.0
   Compiling version_check v0.9.4
   Compiling proc-macro2 v1.0.46
   Compiling cc v1.0.73
   Compiling pkg-config v0.3.25
   Compiling unicode-ident v1.0.0
   Compiling io-lifetimes v1.0.1
   Compiling serde v1.0.137
   Compiling regex-automata v0.1.10
   Compiling log v0.4.17
   Compiling lazy_static v1.4.0
   Compiling crossbeam-utils v0.8.14
   Compiling rustix v0.36.3
   Compiling generic-array v0.14.5
   Compiling regex-syntax v0.6.26
   Compiling syn v1.0.102
   Compiling once_cell v1.12.0
   Compiling bitflags v1.3.2
   Compiling same-file v1.0.6
   Compiling linux-raw-sys v0.1.3
   Compiling serde_json v1.0.81
   Compiling serde_derive v1.0.137
   Compiling fnv v1.0.7
   Compiling aho-corasick v0.7.18
   Compiling bstr v0.2.17
   Compiling quote v1.0.18
   Compiling lzma-sys v0.1.17
   Compiling walkdir v2.3.2
   Compiling thread_local v1.1.4
   Compiling unicode-width v0.1.9
   Compiling ryu v1.0.10
   Compiling bootstrap v0.0.0 (/home/alona/dev/rust/rust/src/bootstrap)
   Compiling cpufeatures v0.2.5
   Compiling itoa v1.0.2
   Compiling getopts v0.2.21
   Compiling opener v0.5.0
   Compiling object v0.29.0
   Compiling cmake v0.1.48
   Compiling filetime v0.2.16
   Compiling xattr v0.2.3
   Compiling build_helper v0.1.0 (/home/alona/dev/rust/rust/src/tools/build_helper)
   Compiling xz2 v0.1.6
   Compiling block-buffer v0.10.2
   Compiling crypto-common v0.1.3
   Compiling tar v0.4.38
   Compiling hex v0.4.3
   Compiling digest v0.10.3
   Compiling regex v1.5.6
   Compiling sha2 v0.10.2
   Compiling globset v0.4.8
   Compiling fd-lock v3.0.8
   Compiling ignore v0.4.18
   Compiling toml v0.5.9
    Finished dev [unoptimized] target(s) in 18.35s
downloading https://static.rust-lang.org/dist/2023-01-30/rustfmt-nightly-x86_64-unknown-linux-gnu.tar.xz
##################################################################################################################################################### 100.0%
extracting /home/alona/dev/rust/rust/build/cache/2023-01-30/rustfmt-nightly-x86_64-unknown-linux-gnu.tar.xz to /home/alona/dev/rust/rust/build/x86_64-unknown-linux-gnu/rustfmt
downloading https://static.rust-lang.org/dist/2023-01-30/rustc-nightly-x86_64-unknown-linux-gnu.tar.xz
##################################################################################################################################################### 100.0%
extracting /home/alona/dev/rust/rust/build/cache/2023-01-30/rustc-nightly-x86_64-unknown-linux-gnu.tar.xz to /home/alona/dev/rust/rust/build/x86_64-unknown-linux-gnu/rustfmt
downloading https://ci-artifacts.rust-lang.org/rustc-builds/03b01c5bec658081605ab078ad3fbcdb6b30f6c2/rust-std-nightly-x86_64-unknown-linux-gnu.tar.xz
##################################################################################################################################################### 100.0%
extracting /home/alona/dev/rust/rust/build/cache/03b01c5bec658081605ab078ad3fbcdb6b30f6c2/rust-std-nightly-x86_64-unknown-linux-gnu.tar.xz to /home/alona/dev/rust/rust/build/x86_64-unknown-linux-gnu/ci-rustc
downloading https://ci-artifacts.rust-lang.org/rustc-builds/03b01c5bec658081605ab078ad3fbcdb6b30f6c2/rustc-nightly-x86_64-unknown-linux-gnu.tar.xz
##################################################################################################################################################### 100.0%
extracting /home/alona/dev/rust/rust/build/cache/03b01c5bec658081605ab078ad3fbcdb6b30f6c2/rustc-nightly-x86_64-unknown-linux-gnu.tar.xz to /home/alona/dev/rust/rust/build/x86_64-unknown-linux-gnu/ci-rustc
downloading https://ci-artifacts.rust-lang.org/rustc-builds/03b01c5bec658081605ab078ad3fbcdb6b30f6c2/rustc-dev-nightly-x86_64-unknown-linux-gnu.tar.xz
##################################################################################################################################################### 100.0%
extracting /home/alona/dev/rust/rust/build/cache/03b01c5bec658081605ab078ad3fbcdb6b30f6c2/rustc-dev-nightly-x86_64-unknown-linux-gnu.tar.xz to /home/alona/dev/rust/rust/build/x86_64-unknown-linux-gnu/ci-rustc
downloading https://ci-artifacts.rust-lang.org/rustc-builds/03b01c5bec658081605ab078ad3fbcdb6b30f6c2/rust-src-nightly.tar.xz
##################################################################################################################################################### 100.0%
extracting /home/alona/dev/rust/rust/build/cache/03b01c5bec658081605ab078ad3fbcdb6b30f6c2/rust-src-nightly.tar.xz to /home/alona/dev/rust/rust/build/x86_64-unknown-linux-gnu/ci-rustc
Building test helpers
running: "cc" "-O0" "-ffunction-sections" "-fdata-sections" "-fPIC" "-m64" "-o" "/home/alona/dev/rust/rust/build/x86_64-unknown-linux-gnu/native/rust-test-helpers/rust_test_helpers.o" "-c" "/home/alona/dev/rust/rust/tests/auxiliary/rust_test_helpers.c"
exit status: 0
running: ZERO_AR_DATE="1" "ar" "cq" "/home/alona/dev/rust/rust/build/x86_64-unknown-linux-gnu/native/rust-test-helpers/librust_test_helpers.a" "/home/alona/dev/rust/rust/build/x86_64-unknown-linux-gnu/native/rust-test-helpers/rust_test_helpers.o"
exit status: 0
running: "ar" "s" "/home/alona/dev/rust/rust/build/x86_64-unknown-linux-gnu/native/rust-test-helpers/librust_test_helpers.a"
exit status: 0
Uplifting library (stage2 -> stage0)
Error: Unable to find the stamp file /home/alona/dev/rust/rust/build/x86_64-unknown-linux-gnu/stage2-std/x86_64-unknown-linux-gnu/release/.libstd.stamp, did you try to keep a nonexistent build stage?
Build completed unsuccessfully in 0:01:17
```

</details>

CC @jyn514

### PR DESCRIPTION
**Pull request by onur-ozkan**
- check compiler stage before forcing for stage2.
- check if download_rustc is not set before forcing for stage1.

resolves #109286

### CODE REVIEW

**Comment by jyn514** [comment]
I don't have time for reviews, sorry.

r? @Mark-Simulacrum

**Comment by ozkanonur** [comment]
> I don't have time for reviews, sorry.
> 
> r? @Mark-Simulacrum

I think Mark is quite busy too.

Can you review this ? @albertlarsan68

File: src/bootstrap/builder.rs [hunk 1]
```diff
@@ -919,7 +919,7 @@ impl<'a> Builder<'a> {
         host: TargetSelection,
         target: TargetSelection,
     ) -> Compiler {
-        if self.build.force_use_stage2() {
+        if self.build.force_use_stage2(Compiler { stage, host }) {
```
Comment (Mark-Simulacrum) [inline src/bootstrap/builder.rs:922]: I would prefer that we avoid creating Compiler structs if they don't correspond to an already-built compiler (i.e., the only constructor should be `self.compiler`).

Can we refactor both force_use functions to take stage and host as separate values?

**Comment by Mark-Simulacrum** [comment]
r=me with nit fixed

I'm generally not sure this is the right logic - the force-use functions feel pretty iffy to me - but I'm OK patching it for now.

**Comment by ozkanonur** [comment]
@Mark-Simulacrum fixed

**Pull request merged**